ur_controllers/src/passthrough_trajectory_controller.cpp

```diff
@@ -0,0 +1,407 @@
+
+  return config;
+}
+
+controller_interface::CallbackReturn PassthroughTrajectoryController::on_activate(const rclcpp_lifecycle::State& state)
+{
+  number_of_joints_ = joint_names_.size();
+  // clear out vectors in case of restart
+  joint_position_state_interface_.clear();
+  joint_velocity_state_interface_.clear();
+
+  for (uint32_t i = 0; i < state_interface_types_.size() * joint_names_.size(); i++) {
+    if (state_interfaces_[i].get_interface_name() == "position") {
+      joint_position_state_interface_.emplace_back(state_interfaces_[i]);
+    } else if (state_interfaces_[i].get_interface_name() == "velocity") {
+      joint_velocity_state_interface_.emplace_back(state_interfaces_[i]);
+    }
+  }
+
+  return ControllerInterface::on_activate(state);
+}
+
+controller_interface::return_type PassthroughTrajectoryController::update(const rclcpp::Time& time,
+                                                                          const rclcpp::Duration& /* period */)
+{
+  static bool firstpass = true;
+  if (firstpass) {
+    now_ns = time.nanoseconds();
+    firstpass = false;
+  } else {
+    last_time_ns = now_ns;
+    now_ns = time.nanoseconds();
+    period_ns = now_ns - last_time_ns;
+  }
+  /* Keep track of how long the trajectory has been executing, if it takes too long, send a warning. */
+  if (command_interfaces_[CommandInterfaces::PASSTHROUGH_TRAJECTORY_TRANSFER_STATE].get_value() == 4.0) {
+    scaling_factor_ = state_interfaces_[StateInterfaces::SPEED_SCALING_FACTOR].get_value();
+    active_trajectory_elapsed_time_ += static_cast<double>(scaling_factor_ * (period_ns / pow(10, 9)));
+    if (active_trajectory_elapsed_time_ > (max_trajectory_time_ + goal_time_tolerance_.seconds()) &&
+        trajectory_active_) {
+      RCLCPP_WARN(get_node()->get_logger(), "Trajectory should be finished by now. You may want to cancel this goal, "
+                                            "if it is not.");
+      trajectory_active_ = false;
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+rclcpp_action::GoalResponse PassthroughTrajectoryController::goal_received_callback(
+    const rclcpp_action::GoalUUID& /*uuid*/,
+    std::shared_ptr<const control_msgs::action::FollowJointTrajectory::Goal> goal)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Received new trajectory.");
+  // Precondition: Running controller
+  if (get_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
+    RCLCPP_ERROR(get_node()->get_logger(), "Can't accept new trajectories. Controller is not running.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+
+  if (state_interfaces_[StateInterfaces::CONTROLLER_RUNNING].get_value() != 1.0) {
+    RCLCPP_ERROR(get_node()->get_logger(), "Trajectory rejected, controller not running in hardware interface.");
+    return rclcpp_action::GoalResponse::REJECT;
+  }
+
+  if (command_interfaces_[CommandInterfaces::PASSTHROUGH_TRAJECTORY_TRANSFER_STATE].get_value() != 0.0) {
```

Comment:
Accessing interfaces directly from the action server callback thread seems dangerous.